### ISSUE TITLE
Support concatenated messages

### ISSUE THREAD

**Issue by edsiper** [zserge/jsmn#80]
By default JSMN waits until have completed JSON messages, which is OK for most of use cases. But sometimes would be very useful to support concatenated [incomplete] messages, e.g:

``` json
{"key 1": 12345, "key 2": "abc"}{"key 1": 6789,"key 2": "def"}
```

in the above example there are two JSON maps concatenated in the same string, as of now JSMN parse that as OK but will return an error PART if the last one is incomplete:

``` json
{"key 1": 12345, "key 2": "abc"}{"key 1": 6789
```

my goal is to be able to retrieve the first completed message and then be able to continue iterating the incomplete one. My use case is: reading JSON messages from a serial interface, so messages will arrive in multiple chunks and processing the ones are ready ASAP is ideal.

I found a possible workaround enabling JSMN_PARENT_LINKS and iterating until (parent == -1), but I need to touch the JSMN context so would be ideal to get native support for this _feature_. 

If you have a better workaround please let me know.

**Comment by edsiper** [comment]
ping on this.

**Comment by zlolik** [comment]
@edsiper Why you need brackets around key1 and key2? Why not using 

``` c
"key 1": 12345, "key 2": "abc", "key 1": 6789
```

**Comment by pt300** [comment]
@zlolik that wouldn't really work for him as he said messages are retrieved through serial, so I guess the message can be unexpectedly end in any place, even in middle of key or value.
After thinking for a bit I found a way to kind of solve @edsiper 's problem.
Make jsmn stop parsing after hitting end of top level object or array. That way you can use that data and later remove it from the buffer, refill it from serial and parse again. And in case of jsmn reporting ERROR_PART just append more data and retry parsing.

**Comment by rustyrussell** [comment]
Indeed, for JSON RPC it's standard to read into a buffer, then call jsmn_parse() (again).  You really want the first object, but if part of the second object is in the buffer you get JSMN_ERROR_PART.  Worst case this continues with part of the third object, etc. and jsmn_parse() never succeeds.

What's really wanted is a jsmn_parse_single() which only parses up to the end of the first complete token.

The hacky solution is to call jsmn_parse(), ignore JSMN_ERROR_PART, the grovel through tokens[0] and figure out how the span of tokens it covers, and pretend jsmn_parse() returned that.  That's horribly inefficient :(

**Comment by rustyrussell** [comment]
Indeed, for JSON RPC it's standard to read into a buffer, then call jsmn_parse() (again).  You really want the first object, but if part of the second object is in the buffer you get JSMN_ERROR_PART.  Worst case this continues with part of the third object, etc. and jsmn_parse() never succeeds.

What's really wanted is a jsmn_parse_single() which only parses up to the end of the first complete token.  I will be implementing this in a fork, I think.

The hacky solution is to call jsmn_parse(), ignore JSMN_ERROR_PART, the grovel through tokens[0] and figure out how the span of tokens it covers, and pretend jsmn_parse() returned that.  That's horribly inefficient :(
